runtime/src/master.rs

```diff
@@ -0,0 +1,695 @@
+        <T as Trait>::Call
+    {
+        /// A proposal succeeded and was executed. The dids listed are the members whose votes were
+        /// used as proof of authorization. The hash of the proposal is provided.
+        Executed(Vec<Did>, Box<Call>),
+        /// The membership of Master has changed.
+        UnderNewOwnership,
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = MasterError<T>;
+
+        fn deposit_event() = default;
+
+        /// Execute a proposal that has received enough votes. The proposal is a serialized Call.
+        /// This function can be called by anyone, even someone who is not a member of Master.
+        ///
+        /// After a sucessful execution, the current round of votes is cleared and round number is
+        /// increased.
+        #[weight = 10_000 + proposal.get_dispatch_info().weight + 1_000 * auth.len() as u64]
+        pub fn execute(
+            origin,
+            proposal: Box<<T as Trait>::Call>,
+            auth: PMAuth,
+        ) -> DispatchResult {
+            Module::<T>::execute_(origin, proposal, auth)
+        }
+
+        /// Root-only. Sets the members and vote requirement for master. Increases the round number
+        /// and removes the votes for the prevous round.
+        ///
+        /// ```
+        /// # use dock_testnet_runtime::master::Membership;
+        /// # extern crate alloc;
+        /// # use alloc::collections::BTreeSet;
+        /// #
+        /// // Setting the following membership will effectively dissolve the master account.
+        /// Membership {
+        ///     members: BTreeSet::new(),
+        ///     vote_requirement: 1,
+        /// };
+        /// ```
+        ///
+        /// Setting the vote requirement to zero grants unrestricted root access to
+        /// all accounts. It is not recomended to set the vote requirement to zero on a
+        /// production chain.
+        #[weight = 10_000 + 1_000 * membership.members.len() as u64]
+        pub fn set_members(
+            origin,
+            membership: Membership,
+        ) -> DispatchResult {
+            Module::<T>::set_members_(origin, membership)
+        }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    pub fn execute_(
+        origin: T::Origin,
+        proposal: Box<<T as Trait>::Call>,
+        auth: PMAuth,
+    ) -> DispatchResult {
+        ensure_signed(origin)?;
+
+        // check
+        let membership = Members::get();
+        let payload = StateChange::MasterVote(Vote {
+            proposal: proposal.encode(),
+            round_no: Round::get(),
+        })
+        .encode();
+        ensure!(
+            auth.len() as u64 >= membership.vote_requirement,
+            MasterError::<T>::InsufficientVotes,
+        );
+        ensure!(
+            auth.keys().all(|k| membership.members.contains(k)),
+            MasterError::<T>::NotMember,
+        );
+        for (did, sig) in auth.iter() {
```

Comment:
You don't need to verify all signatures, loop can be short-circuited on finding `vote_requirement` no of valid signatures. In fact the check in above ensure can be moved inside this loop if you want to be more efficient.

Reply:
`execute()` is a cold code path. I expect it to be run less than a million times through the life of the chain. You think its worth optimizing?

Reply:
Practically speaking, no.